crates/sui-core/src/quorum_driver/reconfig_observer.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_trait::async_trait;
+use std::sync::Arc;
+use sui_types::committee::Committee;
+use tracing::{info, warn};
+
+use crate::{
+    authority::AuthorityStore,
+    authority_aggregator::{AuthAggMetrics, AuthorityAggregator},
+    authority_client::NetworkAuthorityClient,
+    epoch::committee_store::CommitteeStore,
+    safe_client::SafeClientMetricsBase,
+};
+
+use super::QuorumDriver;
+
+#[async_trait]
+pub trait ReconfigObserver<A> {
+    async fn run(&mut self, quorum_driver: Arc<QuorumDriver<A>>);
+}
+
+pub struct OnsiteReconfigObserver {
+    reconfig_rx: tokio::sync::broadcast::Receiver<Committee>,
+    authority_store: Arc<AuthorityStore>,
+    committee_store: Arc<CommitteeStore>,
+    safe_client_metrics_base: SafeClientMetricsBase,
+    auth_agg_metrics: AuthAggMetrics,
+}
+
+impl OnsiteReconfigObserver {
+    pub fn new(
+        reconfig_rx: tokio::sync::broadcast::Receiver<Committee>,
+        authority_store: Arc<AuthorityStore>,
+        committee_store: Arc<CommitteeStore>,
+        safe_client_metrics_base: SafeClientMetricsBase,
+        auth_agg_metrics: AuthAggMetrics,
+    ) -> Self {
+        Self {
+            reconfig_rx,
+            authority_store,
+            committee_store,
+            safe_client_metrics_base,
+            auth_agg_metrics,
+        }
+    }
+
+    async fn create_authority_aggregator_from_system_state(
+        &self,
+    ) -> AuthorityAggregator<NetworkAuthorityClient> {
+        AuthorityAggregator::new_from_system_state(
+            &self.authority_store,
+            &self.committee_store,
+            self.safe_client_metrics_base.clone(),
+            self.auth_agg_metrics.clone(),
+        )
+        // TODO: we should tolerate when <= f validators give invalid addresses
+        // GH issue: https://github.com/MystenLabs/sui/issues/7019
+        .unwrap_or_else(|e| {
+            warn!("panicked? {:?}", e);
+            panic!(
+                "Failed to create AuthorityAggregator from System State: {:?}",
+                e
+            )
```

Comment:
Why do we need the warning if we have the panic?

Reply:
this is a debug log, will remove. thanks